File: java/pgcon1/Q5.java
package pgcon1;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Q5 {
	static boolean bElapsed = true;
	Map<String,Node> map = new HashMap<>();
	String answerPath = null;
	int answerCost = 0;

	void solve(BufferedReader br) throws Exception {
		for (char ch='A'; ch<='Z'; ch++) {
			map.put(String.valueOf(ch), new Node());
		}
		String line = br.readLine();
		int job = Integer.parseInt(line);
    	for (int i=0; i<job; i++) {
    		line = br.readLine();
    		String[] flds = line.split(" ");
    		String cur = flds[0];
    		int v = Integer.parseInt(flds[1]);
    		int n = Integer.parseInt(flds[2]);
			Node node = map.get(cur);
			node.cost = v;
    		if (cur.equals("A")) {
    			continue;
    		}
    		for (int j=0; j<n; j++) {
    			String prev = flds[3+j];
    			Node prevNode = map.get(prev);
    			prevNode.next.add(cur);
    		}
    	}
    	search("", 0, "A");
    	pln(answerPath);
	}
	
	void search(String path, int cost, String id) {
		if (path.indexOf(id) >= 0) return;
		Node node = map.get(id);
		path += id;
		cost += node.cost;
		check(path, cost);
		for (String s : node.next) {
			search(path, cost, s);
		}
	}
	
	void check(String path, int cost) {
		if (cost > answerCost) {
			answerCost = cost;
			answerPath = path;
		} else if (cost == answerCost) {
			if (path.compareTo(answerPath) < 0) {
				answerPath = path;
			}
		}
	}

	class Node {
		List<String> next = new ArrayList<>();
		int cost;
	}

	void p(char c) {
		System.out.print(c);
	}
	void pln(char c) {
		System.out.println(c);
	}
	void p(long l) {
		System.out.print(l);
	}
	void pln(long l) {
		System.out.println(l);
	}
	void p(String s) {
		System.out.print(s);
	}
	void pln(String s) {
		System.out.println(s);
	}
	//Integer.MAX_VALUE=2147483647>10^9
	//Long.MAX_VALUE=9223372036854775807L>10^18
	public static void main(String[] args) throws Exception {
		long start = System.currentTimeMillis();
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		new Q5().solve(br);
		long end = System.currentTimeMillis();
		if (bElapsed) {
			System.out.println((end-start) + "ms");
		}
	}
}

File: c/pgcon1/q5.c
//プログラムコンテスト
//C言語用
//ファイル名 q5.c
//作成者 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LEN ( 100 )

typedef struct tagNode {
	int cost;
	char next[27];
} Node;

Node nodes[26];
char answerPath[27];
int answerCost;
char line[MAX_LEN];
char *cols[MAX_LEN];

int split(char *str, const char *delim, char *outlist[]) {
    char *token;
    int count = 0;

    token = strtok(str, delim);
    while (token != NULL && count < MAX_LEN) {
        outlist[count++] = token;
        token = strtok(NULL, delim);
    }
    return count;
}

void addPrev(char cur, char prev) {
	char *next;

	next = nodes[prev-'A'].next;
	while (*next) next++;
	*next = cur;
}

void check(char *path, int cost) {
	if (cost > answerCost) {
		answerCost = cost;
		strcpy(answerPath, path);
	} else if (cost == answerCost) {
		if (strcmp(path, answerPath) < 0) {
			strcpy(answerPath, path);
		}
	}
}

void seek(char cur, char *path, int cost) {
    char buf[27];
    char *ptr;
    
    for (ptr=path; *ptr; ptr++) {
        if (*ptr == cur) return;
    }
    strcpy(buf, path);
    for (ptr=buf; *ptr; ptr++) ;
    *ptr++ = cur;
    *ptr = '\0';
    cost += nodes[cur-'A'].cost;
    //printf("%s,%d\n", buf, cost);
    check(buf, cost);
    for (ptr=nodes[cur-'A'].next; *ptr; ptr++) {
        seek(*ptr, buf, cost);
    }
}

int main() {
	int job, i, j, count, v, n;
	char cur, prev;

	fgets(line, sizeof(line), stdin);
	job = atoi(line);
	for (i=0; i<job; i++) {
		fgets(line, sizeof(line), stdin);
		count = split(line, " ", cols);
		cur = cols[0][0];
		v = atoi(cols[1]);
		n = atoi(cols[2]);
		nodes[cur-'A'].cost = v;
		if (cur == 'A') continue;
		for (j=0; j<n; j++) {
			prev = cols[3+j][0];
			addPrev(cur, prev);
		}
	}
    seek('A', "", 0);
    printf("%s\n", answerPath);
    return 0;
}
